File: frontend/lib/core/models/obd.dart
part of models;

class ObdCode {
  final String id, code;
  final String? description;
  final bool isResolved;
  final String source;

  const ObdCode({
    required this.id,
    required this.code,
    this.description,
    this.isResolved = false,
    this.source = 'obd',
  });

  factory ObdCode.fromJson(Map<String, dynamic> j) => ObdCode(
        id: j['id'] as String,
        code: j['code'] as String,
        description: j['description'] as String?,
        isResolved: (j['is_resolved'] as bool?) ?? false,
        source: j['source'] as String? ?? 'obd',
      );
}

File: backend/app/models/obd.py
"""OBD-II fault codes captured from a Bluetooth OBD2 adapter."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.session import Base


def _uuid() -> str:
    return str(uuid.uuid4())


class ObdCode(Base):
    __tablename__ = "obd_codes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    vehicle_id: Mapped[str] = mapped_column(String(36), ForeignKey("vehicles.id"), index=True)
    code: Mapped[str] = mapped_column(String(16), index=True)  # e.g. P0301
    description: Mapped[str | None] = mapped_column(Text)
    source: Mapped[str] = mapped_column(String(20), default="obd")  # obd/manual
    is_resolved: Mapped[bool] = mapped_column(default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
